pages/react-canvas-components.mdx

```diff
@@ -0,0 +1,256 @@
+import { RandomHexagons } from '../components/hexagons/RandomHexagons';
+import { EditOnCodeSandbox } from '../components/EditOnCodeSandbox';
+import { css } from '@emotion/core';
+
+export const meta = {
+  title: 'React + Canvas = 💜',
+  description: 'an untold love story',
+  date: 'January 12, 2019',
+};
+
+The `<canvas/>` HTML element can be used to draw graphics with a finer control than the usual DOM or SVG. But with React, trying to draw on a canvas is not intuitive as their interfaces are quite different. With React, each component own their node contrary to canvas where there is only one shared node, and we draw in it. Let's see how we can make a canvas visualization with React components !
+
+## 🧑‍🏫 Canvas 101
+
+The `canvas` element is like a sheet of paper. To draw in real life, you would take a pen, move your hand to a first position, and draw a line by moving your hand to another position. The browser API to draw on a canvas is actually very similar. We first need to make a blueprint of the shape we want to draw (like using a wood pencil), and then we can color it.
+
+```js
+// moving our hand to the starting position
+canvasContext.moveTo(x1, y1);
+// drawing a blueprint line to the finishing position
+canvasContext.lineTo(x2, y2);
+// taking a purple pen and coloring the line
+canvasContext.strokeStyle = 'purple';
+canvasContext.stroke();
+```
+
+Having imperative code like this in a component oriented codebase can be tricky ! Let's create a component that renders a `<canvas />` on the page. Then, we need to call the `moveTo` and `lineTo` methods on it to draw a line. In practice, it's a bit more complicated to bridge those two. We need to use a [React reference](https://thibaut.io/react-refs-evolution) to access the canvas DOM node, then need to get a 2D context out of it, and finally we can call our drawing methods. The code looks like that:
```

Comment:
- "component-oriented" (with the dash) and no space before the exclamation mark
- " Let's create a component that renders a `<canvas />` on the page. Then, we need to call the `moveTo` and `lineTo` methods on it to draw a line." ==> "We would need to create a component that renders a `<canvas />` on the page and then call the `moveTo` and `lineTo` methods on it to draw a line."
- "to access the canvas DOM node, then need to get a 2D context out of it ... and finally" => "to access the canvas DOM node and to retrieve a 2D context from it; we are then able to call our drawing methods"
- "the code would look something like this:"